.github/workflows/nightly-pypi-build.yml

```diff
@@ -0,0 +1,89 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+
+name: "Nightly PyPI Build"
+
+on:
+  schedule:
+    - cron: "0 0 * * *"  # Runs at midnight UTC every day
+  workflow_dispatch:  # Allows manual triggering
+
+jobs:
+  set-version:
+    if: github.repository == 'apache/iceberg-python'  # Only run for apache repo
+    runs-on: ubuntu-latest
+    outputs:
+      VERSION: ${{ steps.set-version.outputs.VERSION }}
+      RC: ${{ steps.set-version.outputs.RC }}
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 1
+
+      - uses: actions/setup-python@v5
+        with:
+          python-version: 3.12
+
+      - name: Install Poetry
+        run: make install-poetry
+
+      - name: Extract version and rc
+        id: set-version
+        run: |
+          VERSION=$(poetry version --short)
+          RC="$(date +%Y%m%d)"
+          echo "VERSION=$VERSION" >> "$GITHUB_OUTPUT"
+          echo "RC=$RC" >> "$GITHUB_OUTPUT"
+
+      - name: Debug version and rc
+        run: echo "Publishing version ${{ steps.set-version.outputs.VERSION }}rc${{ steps.set-version.outputs.RC }}"
+
+  nightly-build:
+    needs: set-version
+    uses: ./.github/workflows/pypi-build-artifacts.yml
+    with:
+      version: ${{ needs.set-version.outputs.VERSION }}  # i.e. 0.9.0
+      rc: ${{ needs.set-version.outputs.RC }}  # i.e. 20250203
+
+  testpypi-publish:
+    name: Publish to TestPypi
+    needs:
+      - nightly-build
+    runs-on: ubuntu-latest
+    environment:
+      name: testpypi
+      url: https://test.pypi.org/p/pyiceberg
+
+    permissions:
+      id-token: write  # IMPORTANT: mandatory for trusted publishing
+
+    steps:
+    - name: Download all the artifacts
+      uses: actions/download-artifact@v4
+      with:
+        merge-multiple: true
+        path: dist/
+    - name: List downloaded artifacts
+      run: ls -R dist/
```

Comment:
I was hoping that we could leave out this kind of debug statements. But it looks like the merge job doesn't mention the actual files: https://github.com/kevinjqliu/iceberg-python/actions/runs/13119881319/job/36603748563